dm/simulator/internal/sqlgen/impl.go

```diff
@@ -0,0 +1,255 @@
+// limitations under the License.
+
+package sqlgen
+
+import (
+	"strings"
+
+	"github.com/chaos-mesh/go-sqlsmith/util"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/format"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/parser/opcode"
+	_ "github.com/pingcap/tidb/types/parser_driver" // import this to make the parser work
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/simulator/internal/config"
+	"github.com/pingcap/tiflow/dm/simulator/internal/mcp"
+)
+
+type sqlGeneratorImpl struct {
+	tableConfig *config.TableConfig
+	columnMap   map[string]*config.ColumnDefinition
+	ukMap       map[string]struct{}
+}
+
+// NewSQLGeneratorImpl generates a new implementation object for SQL generator.
+func NewSQLGeneratorImpl(tableConfig *config.TableConfig) *sqlGeneratorImpl {
+	colDefMap := make(map[string]*config.ColumnDefinition)
+	for _, colDef := range tableConfig.Columns {
+		colDefMap[colDef.ColumnName] = colDef
+	}
+	ukMap := make(map[string]struct{})
+	for _, ukColName := range tableConfig.UniqueKeyColumnNames {
+		if _, ok := colDefMap[ukColName]; ok {
+			ukMap[ukColName] = struct{}{}
+		}
+	}
+	return &sqlGeneratorImpl{
+		tableConfig: tableConfig,
+		columnMap:   colDefMap,
+		ukMap:       ukMap,
+	}
+}
+
+// outputString parses an ast node to SQL string.
+func outputString(node ast.Node) (string, error) {
+	var sb strings.Builder
+	err := node.Restore(format.NewRestoreCtx(format.DefaultRestoreFlags, &sb))
+	if err != nil {
+		return "", errors.Annotate(err, "restore AST into SQL string error")
+	}
+	return sb.String(), nil
+}
+
+// GenTruncateTable generates a TRUNCATE TABLE SQL.
+// It implements the SQLGenerator interface.
+func (g *sqlGeneratorImpl) GenTruncateTable() (string, error) {
+	truncateTree := &ast.TruncateTableStmt{
+		Table: &ast.TableName{
+			Schema: model.NewCIStr(g.tableConfig.DatabaseName),
+			Name:   model.NewCIStr(g.tableConfig.TableName),
+		},
+	}
+	return outputString(truncateTree)
+}
+
+func (g *sqlGeneratorImpl) generateWhereClause(theUK map[string]interface{}) (ast.ExprNode, error) {
+	compareExprs := make([]*ast.BinaryOperationExpr, 0)
+	// iterate the existing UKs, to make sure all the uk columns has values
+	for ukColName := range g.ukMap {
+		val, ok := theUK[ukColName]
+		if !ok {
+			log.L().Error(ErrUKColValueNotProvided.Error(), zap.String("column_name", ukColName))
+			return nil, errors.Trace(ErrUKColValueNotProvided)
+		}
+		compareExprs = append(compareExprs, &ast.BinaryOperationExpr{
+			Op: opcode.EQ,
+			L: &ast.ColumnNameExpr{
+				Name: &ast.ColumnName{
+					Name: model.NewCIStr(ukColName),
+				},
+			},
+			R: ast.NewValueExpr(val, "", ""),
+		})
+	}
+	return generateCompoundBinaryOpExpr(compareExprs), nil
+}
+
+func generateCompoundBinaryOpExpr(compExprs []*ast.BinaryOperationExpr) ast.ExprNode {
+	switch len(compExprs) {
+	case 0:
```

Comment:
Do we support using a empty UniqueKeyColumnNames?

For other cases this will cause ill-formed "AND" Op, we should panic or log some error